flink-state-backends/flink-statebackend-changelog/src/main/java/org/apache/flink/state/changelog/PeriodicMaterializationManager.java

```diff
@@ -104,7 +105,7 @@ public void start() {
 
             LOG.info("Task {} starts periodic materialization", subtaskName);
 
-            scheduleNextMaterialization();
+            scheduleNextMaterialization(new Random().nextInt((int) periodicMaterializeDelay));
```

Comment:
Materializations should be spread evenly across operators, not only subtasks. And while it is possible to use operator ID and e.g. compute hash from it, the total number of operators is unknown. So I don't see how this (hashcode+subtaskIndex) could be transformed into a number with equal probability in range `[0..delay)`. Do you have any ideas?

Besides that:
- operator ID itself is generated randomly (so re-submitted a job will have different order)
- issues with Random number generation will likely affect operator ID too, as well as many other places
- hashing can be thought of as a form of randomization; distribution quality depends on the hash function, instead of RND; which is better in most cases, but not all

Reply:
Isn't the initialisation order of the state backends per operator within a subtask deterministic? So wouldn't be a hash of `env.getTaskInfo().getTaskName() + env.getTaskInfo().getIndexOfThisSubtask() + incrementalId` enough?

Reply:
Yes, the order is deterministic, but I don't see how this can help.
The hash code cab be arbitrarily large, so it needs to be transformed into range `[0..delay)`. For that, the number of operators is needed.
Known is the number of operators in each chain. But it can be the same in different chains, so the delay would be the same.
The number of chains, as well as the total number of operators is unknown on TM AFAIK (at least without involving JM).

Reply:
Ok, I see.

Are scheduling periodic materialisation at fixed "wall clock" rate? Or once every N checkpoints?

Reply:
Materializations are scheduled with a fixed interval according to "wall clock". 

In particular, once materialization finishes, it schedules the next one with a fixed delay. This delay is the same for all subtasks/backends. This PR only adds an initial offset to this delay.

Reply:
Anyway, can not we use here:
```
Objects.hash(
  env.getTaskInfo().getTaskName(), 
  env.getTaskInfo().getIndexOfThisSubtask(), 
  incrementalId++) % periodicMaterializeDelay
```?